features/configure-users/configure.d/configure-users

```diff
@@ -0,0 +1,176 @@
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+# Sets credentials for connecting to S3
+_set_s3_configuration() {
+  files_load_config cluster-customizer
+  _S3CFG="$(mktemp /tmp/cluster-customizer.s3cfg.XXXXXXXX)"
+  cat <<EOF > "${_S3CFG}"
+[default]
+access_key = "${cw_CLUSTER_CUSTOMIZER_access_key_id}"
+secret_key = "${cw_CLUSTER_CUSTOMIZER_secret_access_key}"
+security_token = ""
+use_https = True
+check_ssl_certificate = True
+EOF
+}
+
+
+# Pulls the user and groups list from client's S3 bucket
+_retrieve_list_files() {
+  USER_FILE="$(mktemp /tmp/cluster-customizer.user-list.XXXXXXXX)"
+  GROUP_FILE="$(mktemp /tmp/cluster-customizer.group-list.XXXXXXXX)"
+  if [ -n "${cw_CLUSTER_CUSTOMIZER_bucket}" ]; then
+    _pull_files "${cw_CLUSTER_CUSTOMIZER_bucket}"
+  elif network_is_ec2; then
+    _pull_files "s3://alces-flight-$(network_ec2_hashed_account)"
+  else
+    echo "ERROR: Clusterware s3 bucket has not been set" >&2
+    exit 1
+  fi
+}
+
+_pull_files() {
+  local bucket
+  bucket=$1
+  out=$(${cw_ROOT}/opt/s3cmd/s3cmd "-c ${_S3CFG}" --force get $bucket/configure-users/userlist.cfg $USER_FILE 2>&1 >/dev/null)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not locate user list: $out" >&2
+    exit 1
+  fi
+  out=$(${cw_ROOT}/opt/s3cmd/s3cmd "-c ${_S3CFG}" --force get $bucket/configure-users/grouplist.cfg $GROUP_FILE 2>&1 >/dev/null)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not locate group list: $out" >&2
+    exit 1
+  fi
+}
+
+# Reads the groups from the list
+_add_groups_from_list() {
+  delim="\s*,\s*"
+  while read -r line || [[ -n "$line" ]]; do
+    #Validates input
+    regex="^[a-z][-a-z0-9]*$delim[0-9]+$"
+    if [[ "$line" =~ $regex ]]; then
+      _add_group $(echo $line | sed "s/$delim/ /g")
+    elif [[ "$line" ]] && ! [[ "$line" =~ ^# ]]; then
+      echo "ERROR: Unrecognised group format: $line" 1>&2
+      exit 1
+    fi
+  done < $GROUP_FILE
+}
+
+# Reads the users from the list
+_add_users_from_list() {
+  delim="\s*,\s*"
+  while read -r line || [[ -n "$line" ]]; do
+    # Validates input
+    regex="^[a-z][-a-z0-9]*($delim[0-9]+)+$"
+    if [[ "$line" =~ $regex ]]; then
+      _add_user $(echo $line | sed "s/$delim/ /g")
+    elif [[ "$line" ]] && ! [[ "$line" =~ ^# ]]; then
+      echo "ERROR: Unrecognised user format: $line" 1>&2
+      exit 1
+    fi
+  done < $USER_FILE
+}
+
+# Adds group ($1 groupname, $2 gid)
+_add_group() {
+  local groupname gid
+  groupname=$1
+  gid=$2
+
+  out=$(groupadd -g $gid $groupname 2>&1)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not add group $gid, $out" >&2
+    exit 1
+  fi
+}
+
+# Adds user ($1 username, $2 uid, $3 gid, $4+ secondary gids)
+_add_user() {
+  local user uid gid
+  username=$1
+  uid=$2
+  gid=$3
+  shift 3
+
+  # Adds the user (and creates primary group if it doesn't exists)
+  if [[ -z "$gid" ]]; then
+    getent group $uid >/dev/null || _add_group $username $uid
+    out=$(useradd -u $uid -g $uid)
```

Comment:
Yup, fair enough.  If the group is implicitly created it will be allocated the next available GID which may not be the same on all the nodes (likely to be, but not guaranteed), so your approach is sound.